fhir-bulkimportexport-webapp/src/main/java/com/ibm/fhir/jbatch/bulkdata/export/fast/ResourcePayloadReader.java

```diff
@@ -0,0 +1,683 @@
+    String cosCredentialIbm;
+
+    /**
+     * Fhir export format.
+     */
+    @Inject
+    @BatchProperty(name = Constants.EXPORT_FHIR_FORMAT)
+    protected String fhirExportFormat;
+
+    /**
+     * Fhir Search from date.
+     */
+    @Inject
+    @BatchProperty(name = Constants.EXPORT_FHIR_SEARCH_FROMDATE)
+    String fhirSearchFromDate;
+
+    /**
+     * Fhir search to date.
+     */
+    @Inject
+    @BatchProperty(name = Constants.EXPORT_FHIR_SEARCH_TODATE)
+    String fhirSearchToDate;
+
+    /**
+     * Fhir export type filters. Ignored for the basicsystem export
+     */
+    @Inject
+    @BatchProperty(name = Constants.EXPORT_FHIR_SEARCH_TYPEFILTERS)
+    String fhirTypeFilters;
+
+    /**
+     * Fhir search page size. Not used
+     */
+    @Inject
+    @BatchProperty(name = Constants.EXPORT_FHIR_SEARCH_PAGESIZE)
+    String fhirSearchPageSize;
+
+    @Inject
+    @BatchProperty(name = Constants.INCOMING_URL)
+    String incomingUrl;
+
+    // Control the number of records to read in each "item".
+    @Inject
+    @BatchProperty(name = Constants.COS_BUCKET_FILE_MAX_RESOURCES)
+    String cosBucketFileMaxResources;
+    int resourcesPerItem = Constants.DEFAULT_MAX_RESOURCES_PER_ITEM;
+
+    // The Java Batch context object
+    @Inject
+    StepContext stepCtx;
+
+    // The context object representing this export job
+    @Inject
+    JobContext jobContext;
+
+    // Track the resources we've processed sharing the most recent timestamp - for correct checkpoint
+    private Set<Long> resourcesForLastTimestamp = new HashSet<>();
+
+    // Each time the timestamp changes, we need to clear resourcesForLastTimestamp
+    private Instant lastTimestamp;
+
+    // The last_modified timestamp to start scanning from
+    private Instant fromLastModified;
+
+    // An end time to scan to, if any
+    private Instant toLastModified;
+
+    // Cap the part upload size to avoid local memory issues. Also need to avoid transaction timeout
+    private long partUploadTriggerSize = Constants.COS_PART_MINIMALSIZE * 10L;
+
+    // How large should a single COS item (file) be
+    private long maxItemSize = Constants.DEFAULT_COSFILE_MAX_SIZE;
+
+    // The initial size of the buffer we use for export. Most resources are
+    // under 10K so this is a reasonable initial value
+    private final int initialBufferSize = 10 * 1024;
+
+    private static final char NDJSON_LINE_SEPARATOR = '\n';
+
+    // Assume for now we have a transaction timeout of 120s, but need to give the upload some time to work
+    // TODO configuration
+    private long txTimeoutMillis = 90000L;
+
+    // The nanoTime after which we want to stop processing to avoid a Liberty transaction timeout
+    private long txEndTime;
+
+    // The buffer space we collect data into until we hit the threshold to push to COS
+    private ByteArrayOutputStream outputStream;
+
+    // The sum of the multi-part sizes for parts that have been uploaded
+    private long currentItemSize;
```

Comment:
same question as above about whether this is 1:1 with java-batch items

Reply:
In this implementation we don't return actual items from the Java-batch ItemReader. I'll work on renaming.